### PR TITLE
Try to restart httpd on Fedora if config check fails

### PR DESCRIPTION
**Pull request by joohoi**
This PR adds a step to Apache plugin config_test when run on Fedora. Because Fedora now creates self signed certificate and related key material upon first startup of httpd. This was causing issues for users who run `certbot-auto` or install certbot (and mod_ssl) and run Certbot directly after.

Fixes: #6828

### CODE REVIEW

**Comment by bmw** [comment]
To help @adferrand learn more about the Apache plugin, he's going to give this an initial review and I'll take another look afterwards.

File: certbot-apache/certbot_apache/tests/centos_test.py [hunk 1]
```diff
@@ -30,6 +30,58 @@ def get_vh_truth(temp_dir, config_name):
...
+            mock_test.side_effect = errors.MisconfigurationError
+            with mock.patch("certbot.util.get_os_info") as mock_info:
+                mock_info.return_value = ["not_fedora"]
+                self.assertRaises(errors.MisconfigurationError,
+                                  self.config.config_test)
+
+    def test_fedora_restart_error(self):
+        c_test = "certbot_apache.configurator.ApacheConfigurator.config_test"
+        with mock.patch(c_test) as mock_test:
+            mock_test.side_effect = [errors.MisconfigurationError, '']
```
Comment (adferrand) [inline certbot-apache/certbot_apache/tests/centos_test.py]: Has noted in the comment below for the same instruction, we should say here that first call raises an error, second does not.

File: certbot-apache/certbot_apache/tests/centos_test.py [hunk 2]
```diff
@@ -30,6 +30,58 @@ def get_vh_truth(temp_dir, config_name):
...
+            with mock.patch("certbot.util.run_script") as mock_run:
+                mock_run.side_effect = errors.SubprocessError
+                self.assertRaises(errors.MisconfigurationError,
+                                  self._run_fedora_test)
+
+    def test_fedora_restart(self):
+        c_test = "certbot_apache.configurator.ApacheConfigurator.config_test"
+        with mock.patch(c_test) as mock_test:
```
Comment (adferrand) [inline certbot-apache/certbot_apache/tests/centos_test.py]: Why `c_test` is mocked two times ?

File: certbot-apache/certbot_apache/override_centos.py [hunk 3]
```diff
@@ -40,6 +43,38 @@ class CentOSConfigurator(configurator.ApacheConfigurator):
...
+            else:
+                raise
+
+    def _try_restart_fedora(self):
+        """
+        Tries to restart httpd using systemctl to generate the self signed keypair.
+        """
+
+        try:
+            util.run_script(['systemctl', 'restart', 'httpd'])
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot-apache/certbot_apache/override_centos.py]: `override_centos` is called for all distros of RHEL family, the list can be seen in [entrypoint.py](https://github.com/certbot/certbot/blob/master/certbot-apache/certbot_apache/entrypoint.py#L12-L28).

This part of the code gets executed only if the "default" `config_test()` raises an exception. If `systemd` isn't installed, this call to `util.run_script` will raise the same exception, so the behavior is the same as previously, and as expected. The "default" config test is run after the restart as well, so compared to the current `master` branch behavior, an alternative code paths are taken only if `systemd restart httpd` fixes the Apache config test error.

File: certbot-apache/certbot_apache/override_centos.py [hunk 4]
```diff
@@ -40,6 +43,38 @@ class CentOSConfigurator(configurator.ApacheConfigurator):
...
+        except errors.SubprocessError as err:
+            raise errors.MisconfigurationError(str(err))
+
+        # Finish with actual config check to see if systemctl restart helped
+        super(CentOSConfigurator, self).config_test()
```
[the remark this answers is not in the document]
Comment (joohoi) [reply certbot-apache/certbot_apache/override_centos.py]: This should be taken care of by the Fedora systemd service already, as the process should not start before the dependency `one-shot` service ([that generates the key material and the cert](https://src.fedoraproject.org/rpms/httpd/pull-request/1#request_diff)) has been run.

File: certbot-apache/certbot_apache/tests/centos_test.py (hunk 2 above)
Comment (joohoi) [reply certbot-apache/certbot_apache/tests/centos_test.py]: I moved things around just before finishing the PR, it's now fixed.

File: certbot-apache/certbot_apache/override_centos.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (adferrand) [reply certbot-apache/certbot_apache/override_centos.py]: Oh right, it is properly registered in the chain of events in systemd (would be expected, by the way ^^)

File: certbot-apache/certbot_apache/override_centos.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (adferrand) [reply certbot-apache/certbot_apache/override_centos.py]: Ok, so the only corner case would be a recent Fedora, with an insane user that managed to uninstall systemd without breaking its system. If it is that level, he will accommodate with some inconsistencies in Certbot behavior ^^

[the remark this answers is not in the document]
Comment (joohoi) [reply certbot-apache/certbot_apache/override_centos.py]: Yeah, indeed :) On top of that they'd have to have a broken Apache configuration as well.

**Pull request merged**